torch/distributed/_functional_collectives.py

```diff
@@ -1041,6 +1041,20 @@ def _reduce_scatter_tensor_coalesced_native_meta(
     )
 
 
+# See Note [Recomputing collectives in the partitioner]
+@contextlib.contextmanager
+def treat_collectives_as_safe_to_dce():
+    torch.fx.node._side_effectful_functions.remove(
+        torch.ops._c10d_functional.wait_tensor.default
+    )
+    try:
+        yield
+    finally:
+        torch.fx.node._side_effectful_functions.add(
+            torch.ops._c10d_functional.wait_tensor.default
+        )
```

Comment:
nesting this context manager will lead to some weird things. That being said you aren't nesting it but it shouldn't be too hard to get it to work? (e.g. only add back what you removed if it was removed)